Add ability to specify path to TF binary

paultyng opened the pull request.

radeksimko commented on langserver/langserver.go:
```diff
@@ -12,12 +12,15 @@ import (
 type langServer struct {
 	srvCtx     context.Context
 	logger     *log.Logger
+	tfPath     discovery.DiscoveryFunc
```
Nitpick, but do you mind calling this `tfDiscoFunc` or something similar that makes it more obvious that this variable is a function and not a string?

I suppose the other `tfPath` in `service.go` is a forgotten relic that should also be changed - especially once https://github.com/hashicorp/terraform-ls/issues/69 is addressed. I'm happy to address it as part of #69 though - no biggie.

radeksimko commented on langserver/session/types.go:
```diff
@@ -5,12 +5,15 @@ import (
 	Assigner() (jrpc2.Assigner, error)
 	Finish(jrpc2.ServerStatus)
 	SetLogger(*log.Logger)
+	SetDiscovery(discovery.DiscoveryFunc)
```
Nitpick: Can we call this `SetDiscoveryFunc`, so it's more obvious that the argument is a function?

radeksimko commented on langserver/handlers/service.go:
```diff
@@ -53,6 +53,10 @@ func (svc *service) SetLogger(logger *log.Logger) {
 	svc.logger = logger
 }
 
+func (svc *service) SetDiscovery(f discovery.DiscoveryFunc) {
```
Nitpick: Can we call this `SetDiscoveryFunc`?

paultyng commented:
@radeksimko made the suggested adjustments, also added some testing of the path string to make sure its a bit more valid as discussed. I cleaned up a few of the other `ui.Error` calls in another commit as well just for consistency. Let me know if its still good to merge.

radeksimko commented on commands/serve_command.go:
```diff
@@ -79,18 +82,43 @@ func (c *ServeCommand) Run(args []string) int {
+		if stat.IsDir() {
+			c.Ui.Error(fmt.Sprintf("Expected a Terraform binary, got a directory: %q", path))
+			return 1
+		}
```
Total nitpick, but I assume the above lines could make a `validateTfExecPath(string) error` function.

paultyng commented:
The confusing part was they weren't on all errors just some errors, but if its the how it should be presented you could already create a fancy UI implementation that adds the separation.

I was hoping to maybe revisit/refactor this after the release anyway, clean up the hard coded exit codes and flag validation a bit. I'll just write an issue to do so if thats ok with you?
